### PR TITLE
[WIP] Add Wayland support

### PR DESCRIPTION
**Pull request by ebkalderon**
### Added
* Introduce new platform-specific define `RENDERDOC_WINDOWING_WAYLAND`.
* Introduce new CMake feature switch `ENABLE_WAYLAND` (set to `OFF` by default).
* Add `FindWayland.cmake` to *scripts/* directory.
* Add `CreateWaylandWindowingData` function to replay API.
* Hook up Wayland support in Vulkan driver.
* Link `librenderdoc.so` and `qrenderdoc` with `libwayland-client.so`.

### Changed
* Append `Wayland` to end of `WindowingSystem` enumeration in replay API.
* Have `qrenderdoc` choose between Wayland or X11 code path based on the active Qt platform being used (can be launched manually with `./qrenderdoc --platform wayland`, requires [qtwayland](https://wiki.qt.io/QtWayland) module installed).
* Enable Qt Wayland support with `qrenderdoc` if `libwayland` is detected on the system.

Fixes issue #853.

### CODE REVIEW

File: qrenderdoc/Code/CaptureContext.cpp
Comment (baldurk) [inline qrenderdoc/Code/CaptureContext.cpp]: The `struct wl_surface *` instead of `wl_surface *` is a C-ism which isn't needed for C++, so I'd rather the `struct` was omitted - the same thing occurs in a few places where these structs are referenced.

File: qrenderdoc/qrenderdoc.pro
Comment (baldurk) [inline qrenderdoc/qrenderdoc.pro]: You'll need to add these defines/libs externally, at the moment this is being added unconditionally even if wayland support was not enabled. This is why the travis build failed until you apt-get'd libwayland-dev - the same thing would happen for an end-user if they didn't have libwayland-dev installed and we want to keep this optional.

If you look in `qrenderdoc/CMakeLists.txt` there's a `qrenderdoc_cmake.pri` include file that's included at the top of `qrenderdoc.pro` and is used to pass cmake configuration to qt. The way PySide2 is linked in only when found is a good example to copy.

File: renderdoc/api/replay/renderdoc_replay.h
Comment (baldurk) [inline renderdoc/api/replay/renderdoc_replay.h]: These comments are formatted for the python bindings first (with being good interface docs for C++ a close second), so the pointer syntax won't necessarily show up well. It's still clear enough if you just put `:param wl_display display` and that will definitely show correctly.

File: renderdoc/os/posix/linux/linux_stringio.cpp
Comment (baldurk) [inline renderdoc/os/posix/linux/linux_stringio.cpp]: I realise this code isn't called yet, but I had a couple of comments on this.

I assume you'll check whether we're running wayland outside this function, because it assumes we are definitely using wayland by not checking if `wl_display_connect()` succeeded.

Also the code as it stands suggests that we don't have to do anything at all to grab keyboard input globally. We connect to the display just with `NULL`, and don't use the window handle at all. It's not necessarily wrong I'm just checking my assumptions that it works like xlib/xcb in that respect. I'm also assuming there's no problem with opening a second wayland display handle here, on xlib/xcb that could cause driver crashes if we don't consistently use only a single connection.

Lastly there's one other issue - in `gl_hooks_egl.cpp` in the hooked `eglGetDisplay` we call `CloneDisplay()` further up in this file on the native display handle passed, assuming it will be an xlib `Display *`. On wayland I suspect that treating a wayland display handle like an xlib display handle will probably crash. Might it be a good idea to move this code that calls `wl_display_connect()` into that function - then if the wayland connection succeeds we assume the display we were just passed was a wayland display and not an xlib display. Does that make sense?

File: scripts/travis/linux_setup.sh
Comment (baldurk) [inline scripts/travis/linux_setup.sh]: This shouldn't be necessary with the qmake change I mentioned above. I'd rather test that the build isn't accidentally relying on wayland, than wayland compilation still works (ideally we'd have a build per configuration but that's probably prohibitive).

**Review by baldurk — CHANGES_REQUESTED** [review]
Thanks, this looks good! Hopefully you were able to find your way around the codebase OK.

I've made some minor comments for changes in a few places. Mostly just stylistic or tweaking to fit better.

One global thing - please rebase or squash to remove any formatting-only commits. Whether you squash entirely or just merge the formatting changes into their respective commits is up to you, but I prefer to keep the history clean wrt formatting.

File: qrenderdoc/Code/CaptureContext.cpp
Comment (ebkalderon) [reply qrenderdoc/Code/CaptureContext.cpp]: Sounds good! I'll reference these structs without the explicit keyword.

File: qrenderdoc/qrenderdoc.pro
Comment (ebkalderon) [reply qrenderdoc/qrenderdoc.pro]: Cool, I'll make this change and revert the `linux_setup.sh` script.

File: renderdoc/api/replay/renderdoc_replay.h
Comment (ebkalderon) [reply renderdoc/api/replay/renderdoc_replay.h]: Ah, I didn't know that. Sure, I'll remove the `struct` keyword and pointer syntax.

File: renderdoc/os/posix/linux/linux_stringio.cpp
Comment (ebkalderon) [reply renderdoc/os/posix/linux/linux_stringio.cpp]: Yeah, this code is certainly not finished. But in response to your questions, yes, we are assuming that we are definitely using Wayland when calling this function since the rest of the codebase is littered with X11-related `ifdef`s and other assumptions, so for the purposes of hacking in basic support, I was making a similar assumption as well.

With Wayland, `wl_display_connect(NULL)` returns a pointer to the currently active `wl_display` and doesn't open a new one. Replacing the `NULL` with a `const char *` will attempt to connect to a different display with the given name. AFAICT, Wayland applications can open as many `wl_display` handles as they want. As for not using `wl_surface` at the moment, the intent is be to eventually pass our `wl_surface` into the keyboard callbacks via `wl_keyboard_set_user_data` to have them verify from which window the events originated before recording them.

Very good point about the `CloneDisplay` issue. I think I'll move this Wayland display connection code to a `CloneDisplay` function like you suggested. But then again, I wrote this code before I added a `wl_display` to `WindowingData`, so it's possible that this code won't even be needed.

File: scripts/travis/linux_setup.sh
Comment (ebkalderon) [reply scripts/travis/linux_setup.sh]: Sounds good. I'll revert this change.

File: renderdoc/os/posix/linux/linux_stringio.cpp
Comment (ebkalderon) [reply renderdoc/os/posix/linux/linux_stringio.cpp]: @baldurk Looks like GitHub decided to hide these comments since I changed this one line from `struct wl_registry *` to just `wl_registry *`. I hope you see this!

**Comment by baldurk** [comment]
I'm not clear on what you're wanting to do with this PR. If you want to land the changes in stages and review each piece then that's fine. However moving the goalposts by adding new code at the same time is difficult to handle. If you want to iterate on your code in a branch and then get it reviewed in the end that's also fine but you'd need to wait to open the PR until you're ready to fix a set of code and then just stabilise/address issues on that fixed set of work until it's merged.

I don't mind either way but you have to pick one or the other :smile:.

File: renderdoc/os/posix/linux/linux_stringio.cpp
Comment (baldurk) [reply renderdoc/os/posix/linux/linux_stringio.cpp]: Yep I see the emails for each comment :smile:.

It sounds like you have a handle on it all, I just wanted to be sure there weren't lurking issues from problems I've run into in the past.

Also note that the `WindowingData` is only used during replay, it isn't part of the code for the injected application at all, and similarly the keyboard input is only used during capture in the injected application for detecting shortcut keypresses. So in that respect it's not related to the code here.

**Comment by ebkalderon** [comment]
You're definitely right, I shouldn't get ahead of myself. I think I prefer the piecemeal approach out of the two listed above, so I think I'll rename this PR, squash the commits, have it reviewed, and hopefully merged.

**Comment by baldurk** [comment]
Just to give a concrete deadline: I won't accept any more PRs for v1.0 after [00:01 UTC on 1st March](https://www.timeanddate.com/countdown/generic?iso=20180301T0001&p0=%3A&font=cursive&csz=1) (so by the end of the day 28th February).

The PR needs to be ready to merge by then if it goes in for v1.0, otherwise it will wait until v1.0 is released and can be merged after that.

**Comment by ebkalderon** [comment]
Thanks for the heads up! Unfortunately, I got sidetracked with real world obligations, and I likely won't be able to finish this in time for the 1.0 release. Nonetheless, I plan to rebase this on 1.0 and resume work on it this weekend.

Congratulations on the incoming release, though. :tada:

**Comment by baldurk** [comment]
I'm going to close this PR now as the changes/discussion is a bit stale - I'd rather not have the PR sticking around indefinitely as that's more what I see issues as being for, long-lived markers for bugs/features and related discussion.

As I say though I'm still happy to get the changes eventually, so the issue will stay open and anyone can chime in there. A new PR (based on these changes or not) would be quite welcome!

**Pull request closed without merging**
